src/storage/src/hummock/compactor/fast_compactor_runner.rs

```diff
@@ -0,0 +1,588 @@
+            task_progress.clone(),
+        ));
+        let right = Box::new(ConcatSstableIterator::new(
+            task.input_ssts[1].table_infos.clone(),
+            context.sstable_store.clone(),
+            task_progress,
+        ));
+
+        Self {
+            executor: CompactTaskExecutor::new(sst_builder, task_config),
+            left,
+            right,
+            task_id: task.task_id,
+            compression_algorithm,
+        }
+    }
+
+    pub async fn run(mut self) -> HummockResult<Vec<SplitTableOutput>> {
+        self.left.rewind().await?;
+        self.right.rewind().await?;
+        let mut skip_raw_block_count = 0;
+        let mut skip_raw_block_size = 0;
+        while self.left.is_valid() && self.right.is_valid() {
+            let ret = self
+                .left
+                .current_sstable()
+                .key()
+                .cmp(&self.right.current_sstable().key());
+            let (first, second) = if ret == Ordering::Less {
+                (&mut self.left, &mut self.right)
+            } else {
+                (&mut self.right, &mut self.left)
+            };
+            assert!(ret != Ordering::Equal);
+            if first.current_sstable().iter.is_none() {
+                let right_key = second.current_sstable().key();
+                while first.current_sstable().is_valid() {
+                    let full_key = FullKey::decode(first.current_sstable().next_block_largest());
+                    // the full key may be either Excluded key or Included key, so we do not allow
+                    // they equals.
+                    if full_key.ge(&right_key) {
+                        break;
+                    }
+                    let smallest_key =
+                        FullKey::decode(first.current_sstable().next_block_smallest());
+                    if self
+                        .executor
+                        .last_key
+                        .user_key
+                        .as_ref()
+                        .eq(&smallest_key.user_key)
+                    {
+                        // If the last key is delete tombstone, we can not append the origin block
+                        // because it would cause a deleted key could be see by user again.
+                        break;
+                    }
+                    let smallest_key =
+                        FullKey::decode(first.current_sstable().next_block_smallest()).to_vec();
+
+                    let (block, filter_data, uncompressed_size) = first
+                        .current_sstable()
+                        .download_next_block()
+                        .await?
+                        .unwrap();
+                    let algorithm = Block::get_algorithm(&block)?;
+                    if algorithm == CompressionAlgorithm::None
+                        && algorithm != self.compression_algorithm
+                        && self.executor.task_config.gc_delete_keys
```

Comment:
This branch needs some comments.
This appears to be a block compacted from above, and it will be gc'd if and only if this condition is met?

Reply:
Yes. We think that if this block has been compact to bottommost level, it does not need to gc again unless it is updated by some other keys from upper level